File: just-bash-shell/repl.ts
/**
 * A tiny REPL that executes bash commands against a Mesa virtual filesystem.
 *
 * This is a minimal REPL and is not a full shell — it's intended as a demonstration
 * of how to interact with a Mesa repo programmatically via just-bash.
 */

import * as readline from 'node:readline';

interface Bash {
  exec(command: string): Promise<{ stdout: string; stderr: string; exitCode: number }>;
}

function question(rl: readline.Interface, prompt: string): Promise<string | null> {
  return new Promise((resolve) => {
    rl.once('close', () => resolve(null));
    rl.question(prompt, (answer) => resolve(answer));
  });
}

/**
 * Spawns an interactive REPL that runs bash commands against a Mesa virtual filesystem.
 */
export default async function tinyBashRepl(bash: Bash): Promise<void> {
  const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout,
  });

  while (true) {
    const input = await question(rl, '$ ');

    if (input === null) {
      break;
    }

    const trimmed = input.trim();
    if (!trimmed) continue;
    if (trimmed === 'exit') break;

    const { stdout, stderr, exitCode } = await bash.exec(trimmed);

    if (stdout) process.stdout.write(stdout);
    if (stderr) process.stderr.write(stderr);

    if (exitCode !== 0) {
      console.error(`[exit ${exitCode}]`);
    }
  }
}

File: py-repl-app-mount/repl.py
"""A tiny REPL that executes bash commands against a Mesa virtual filesystem.

This is a minimal REPL and is not a full shell. It is intended as a
demonstration of how to interact with a Mesa repo programmatically via the
Python SDK's app-level virtual filesystem mount.
"""

from __future__ import annotations

import sys
from collections.abc import Awaitable, Callable

from mesa_sdk import Bash


async def tiny_bash_repl(
    bash: Bash,
    on_finish: Callable[[], Awaitable[None]],
) -> None:
    """Spawn an interactive REPL backed by a Mesa virtual filesystem."""
    while True:
        try:
            line = input("$ ")
        except (EOFError, KeyboardInterrupt):
            print()
            break

        command = line.strip()
        if not command:
            continue
        if command == "exit":
            break

        result = await bash.exec(command)
        await on_finish()

        if result.stdout:
            sys.stdout.buffer.write(result.stdout)
            sys.stdout.buffer.flush()
        if result.stderr:
            sys.stderr.buffer.write(result.stderr)
            sys.stderr.buffer.flush()

        if result.exit_code != 0:
            print(f"[exit {result.exit_code}]", file=sys.stderr)
